tests/integration_test/statistics_test.go

```diff
@@ -0,0 +1,148 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package integration_test
+
+import (
+	"fmt"
+	"testing"
+	"time"
+
+	"github.com/GoogleCloudPlatform/esp-v2/tests/endpoints/echo/client"
+	"github.com/GoogleCloudPlatform/esp-v2/tests/env"
+	"github.com/GoogleCloudPlatform/esp-v2/tests/env/platform"
+	"github.com/GoogleCloudPlatform/esp-v2/tests/utils"
+
+	comp "github.com/GoogleCloudPlatform/esp-v2/tests/env/components"
+)
+
+func roughEqual(i, j, latencyMargin float64) bool {
+	return i > j*(1-latencyMargin) && i < j*(1+latencyMargin)
+}
+
+func TestStatistics(t *testing.T) {
+	t.Parallel()
+
+	s := env.NewTestEnv(comp.TestStatistics, platform.EchoRemote)
+
+	defer s.TearDown()
+	if err := s.Setup(utils.CommonArgs()); err != nil {
+		t.Fatalf("fail to setup test env, %v", err)
+	}
+	const latencyMargin = 0.8
+
+	//the latency is different each run. Here the backend server introduces a
+	// fixed big N second latency, so the overall latency should be around N
+	// second. This test compares the latency with expected N second with certain
+	// error margin. As for the overhead time, the test only checks if the number of
+	// statistic values is correct by setting 0.0, which will be skipped for exact
+	// comparison.
+	testData := []struct {
+		desc           string
+		reqCnt         int
+		reqDuration    time.Duration
+		wantCounts     map[string]int
+		wantHistograms map[string][]float64
+	}{
+
+		{
+			desc:        "backend respond in 1s",
+			reqDuration: time.Second * 1,
+			wantCounts: map[string]int{
+				"http.ingress_http.backend_auth.token_added":                       1,
+				"http.ingress_http.backend_routing.append_path_to_address_request": 1,
+				"http.ingress_http.path_matcher.allowed":                           1,
+				"http.ingress_http.service_control.allowed":                        1,
+			},
+			wantHistograms: map[string][]float64{
+				"http.ingress_http.service_control.overhead_time": {0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0},
+				"http.ingress_http.service_control.backend_time":  {1000, 1025, 1050, 1075, 1090, 1095, 1099, 1099.5, 1099.9, 1100},
+				"http.ingress_http.service_control.request_time":  {1000, 1025, 1050, 1075, 1090, 1095, 1099, 1099.5, 1099.9, 1100},
+			},
+		},
+		{
+			desc:        "backend respond in 2s",
+			reqDuration: time.Second * 2,
+			wantCounts: map[string]int{
+				"http.ingress_http.backend_auth.token_added":                       2,
+				"http.ingress_http.backend_routing.append_path_to_address_request": 2,
+				"http.ingress_http.path_matcher.allowed":                           2,
+				"http.ingress_http.service_control.allowed":                        2,
```

Comment:
So each test case is influenced by the statistics of the previous test case? This seems difficult to test.

Perhaps reset counters between each test? https://www.envoyproxy.io/docs/envoy/latest/operations/admin#post--reset_counters

If that does not reset the histograms, then just have our integration test suite run `Setup` for each test entry?